core/sessions/ldapauth/ldap.go

```diff
@@ -0,0 +1,640 @@
+// server blocking callback, or an error code to pass a OTP
+func (l *ldapAuthenticator) GetUserWebAuthn(email string) ([]sessions.WebAuthn, error) {
+	return []sessions.WebAuthn{}, nil
+}
+
+// CreateSession will forward the session request credentials to the
+// LDAP server, querying for a user + role response if username and
+// password match. The API call is blocking with timeout, so a sufficient timeout
+// should allow the user to respond to potential MFA push notifications
+func (l *ldapAuthenticator) CreateSession(sr sessions.SessionRequest) (string, error) {
+	// Establish ephemeral connection
+	conn, err := l.dialAndConnect()
+	if err != nil {
+		return "", errors.New("Unable to establish connection to LDAP server with provided URL and credentials")
+	}
+	defer conn.Close()
+
+	// Attempt to LDAP Bind with user provided credentials
+	escapedEmail := ldap.EscapeFilter(strings.ToLower(sr.Email))
+	searchBaseDN := fmt.Sprintf("%s=%s,%s,%s", l.config.BaseUserAttr(), escapedEmail, l.config.UsersDn(), l.config.BaseDn())
+	if err := conn.Bind(searchBaseDN, sr.Password); err != nil {
+		l.lggr.Infof("Error binding user authentication request in LDAP Bind: %v", err)
+		return "", errors.New("Unable to log in with LDAP server. Check credentials")
+	}
+
+	// Bind was successful meaning user and credentials are present in LDAP directory
+	// Reuse FindUser functionality to fetch user roles used to create ldap_session entry
+	// with cached user email and role
+	foundUser, err := l.FindUser(escapedEmail)
+	if err != nil {
+		l.lggr.Infof("Successful user login, but error querying for user groups: user: %s, error %v", escapedEmail, err)
+		return "", errors.New("Log in successful, but no assigned groups to assume role")
+	}
+
+	l.lggr.Infof("Successful LDAP login request for user %s - %s", sr.Email, foundUser.Role)
+
+	// Save session, user, and role to database. Given a session ID for future queries, the LDAP server will not be queried
+	// Sessions are set to expire after the duration + creation date elapsed, and are synced on an interval against the upstream
+	// LDAP server
+	session := sessions.NewSession()
+	_, err = l.q.Exec(
+		"INSERT INTO ldap_sessions (id, user_email, user_role, created_at) VALUES ($1, $2, $3, now())",
+		session.ID,
+		strings.ToLower(sr.Email),
+		foundUser.Role,
+	)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to create new session in ldap_sessions table")
+	}
+
+	l.auditLogger.Audit(audit.AuthLoginSuccessNo2FA, map[string]interface{}{"email": sr.Email})
+
+	return session.ID, nil
+}
+
+// ClearNonCurrentSessions removes all ldap_sessions but the id passed in.
+func (l *ldapAuthenticator) ClearNonCurrentSessions(sessionID string) error {
+	_, err := l.q.Exec("DELETE FROM ldap_sessions where id != $1", sessionID)
+	return err
+}
+
+// CreateUser is not supported for read only LDAP
+func (l *ldapAuthenticator) CreateUser(user *sessions.User) error {
+	return ErrNotSupported
```

Comment:
It's an interesting question. What errors do you think this could/would cause for NOPs? My fear would be that it would be confusing, but not that it would break the node.

To try and call these functions while LDAP is enabled is not a valid thing to do because it's trying to change on the node what a user is allowed to do after saying that is externally defined. I think there definitely needs to be good documentation around this but I don't think this could harm the node, but maybe you have an idea for a code path I don't?

Reply:
@obelisk ^ these kinds of errors are what I was referring to. It sounds like it's actually expected though, and we're basically saying: "don't use this to create users, go to the ldap server", which makes sense 👍